### PR TITLE
Rework CI strategy

### LINKED ISSUE #1018: Add Windows testing platform to Travis
Since we [dropped AppVeyor for Windows testing](amethyst/amethyst#839) a while back, we haven't had any test coverage for Windows builds of Amethyst. However, as of a day ago, Travis CI has officially rolled out native Windows support with the Rust toolchain available from the get-go ([relevant blog post](https://blog.travis-ci.com/2018-10-11-windows-early-release?utm_source=reddit&utm_medium=web&utm_campaign=windows_early_release)). I think that we should enable this to avoid regressions on all supported platforms.

The new `windows` environment in Travis uses Bash for scripting, meaning our existing CI scripts should run fine without modifications, in theory. It also exposes [Chocolatey](https://chocolatey.org/) for package management, should we need to install any native dependencies. Some packages like `curl` and `cmake` are pre-installed and don't need to be explicitly fetched ([complete list of pre-installed packages](https://docs.travis-ci.com/user/reference/windows#pre-installed-packages)).

### PR DESCRIPTION
**Pull request by magnonellie**
Alright here it is, just as I'm getting seriously tired of all things CI.

The main new thing with this strategy is that we will only use Travis for bors and use a self-hosted CI service for everything else, most notably PRs. The Travis configuration has been heavily optimized to the point where a single build only takes on average <25 minutes (tested on travis-ci.com; I don't know if there is a difference). Travis also now tests with Windows (fixes #1018) and has all caching removed as it simply doesn't work for Amethyst due to the sizes of test binaries.

The new CI service is run on a platform called [Concourse](https://concourse-ci.org/). Concourse plays very well with containers (except for within Windows) and as such every build is executed within this [docker image](https://hub.docker.com/r/magnonellie/amethyst-dependencies/). The only build performed on PRs is the stable Linux build as it will catch 99.9% of all errors that can occur. If there is a problem with other platforms or with the beta version of Rust, that will later be caught when running bors in Travis. The testing pipeline is also very flexible which means that we can do things like running a check with the latest rustfmt, catching potential formatting errors. Instead of only showing a single check on PRs for the entire test, the statuses are now separated, allowing you to see what failed before even checking out the concourse page.

Checklist for merging:
- [x] A webhook needs to be added to the repo
- [ ] We need to turn off testing PRs with Travis
- [x] The pipeline config needs to updated to work on the main repo
- [x] [@amethyst-bot](https://github.com/amethyst-bot) needs write permission to the repo
- [ ] (optional) We might want to use a better URL than the current one. Optimally, we would use something like ci.amethyst.rs. I have a checklist for everything that needs to be done in case we do this.

### CODE REVIEW

File: .concourse/pr.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+resource_types:
+- name: pull-request
+  type: docker-image
+  source:
+    repository: teliaoss/github-pr-resource
+
+resources:
+- name: pull-request
+  type: pull-request
+  check_every: 24h
```
Comment (Rhuagh) [inline .concourse/pr.yml]: Does this mean that PRs will take up to 24h before being tested?

File: .concourse/pr.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
+---
+resource_types:
+- name: pull-request
+  type: docker-image
+  source:
+    repository: teliaoss/github-pr-resource
```
Comment (LucioFranco) [inline .concourse/pr.yml]: is it possible to lock this down to a tag?

File: .concourse/tasks/linux-book.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+inputs:
+- name: amethyst
+- name: target
+
+run:
+  path: sh
+  dir: amethyst
+  args:
+  - -exc
```
Comment (LucioFranco) [inline .concourse/tasks/linux-book.yml]: tab here? (*stupid yaml*)

**Comment by fhaynes** [comment]
Please note the server this is running on is unmonitored, and nothing is backed up. @magnonellie do we need anything in that regard?

File: .concourse/pr.yml (hunk 1 above)
Comment (Rhuagh) [reply .concourse/pr.yml]: Ah, that's fine then.

File: .concourse/tasks/linux-book.yml (hunk 3 above)
Comment (Rhuagh) [reply .concourse/tasks/linux-book.yml]: Yaml formatting usually do it this way (without tabs for lists).

**Comment by magnonellie** [comment]
Backing up really shouldn't be necessary as long as we have the docker compose config files backed up somewhere. The only thing we'll lose if things go south are the build logs which aren't that important anyway. Monitoring is a different matter that I'm not super familiar with, but I don't think that should be necessary either.

File: .concourse/tasks/linux-book.yml (hunk 3 above)
Comment (LucioFranco) [reply .concourse/tasks/linux-book.yml]: Yaml is disgusting 😄

**Comment by LucioFranco** [comment]
@magnonellie I just want to verify where are the tests for this PR?

**Comment by magnonellie** [comment]
@LucioFranco That's actually a good question. I thought Travis would step in here and build, but I guess not for some reason...

File: .concourse/pr.yml (hunk 2 above)
Comment (magnonellie) [reply .concourse/pr.yml]: Done!

File: .concourse/tasks/linux-book.yml (hunk 3 above)
Comment (magnonellie) [reply .concourse/tasks/linux-book.yml]: I can change that for all yaml here if you want, but it seems the most commonly accepted way of writing yaml is the way I have done it currently.

Comment (fhaynes) [reply .concourse/tasks/linux-book.yml]: I believe tabs are actually not allowed in YAML? Anyway, the yaml formatting looks fine to me. If we want to have a consistent standard, we could just agree on the rules and use an auto formatter as part of the toolchain?

**Comment by fhaynes** [comment]
We can do ServerSpec tests if you want to test the infra. =) @magnonellie I'll set up some basic metrics and some other tools I think you'll like.

**Comment by magnonellie** [comment]
I'm honestly not sure what happened to Windows there... For some reason there are a bunch of compile errors on font-config?

For the record, Windows did work before: https://travis-ci.com/magnonellie/amethyst/jobs/152682750

File: .concourse/tasks/linux-book.yml (hunk 3 above)
Comment (Rhuagh) [reply .concourse/tasks/linux-book.yml]: I say leave it the way it is, it's the most common way of doing it.

**Comment by magnonellie** [comment]
@ebkalderon Is it possible we could arrange using ci.amethyst.rs?

**Comment by jojolepro** [comment]
Do not merge this before monday, thanks

**Comment by magnonellie** [comment]
@torkleyy Why do we need to update the bors config? Bors is still only gonna need to care about Travis.

**Comment by torkleyy** [comment]
Ah sorry, correct. Btw, love your bots Bio ;)

Shall we try building this again?

**Comment by LucioFranco** [comment]
@magnonellie so what is the status on this, I still don't understand why this PR isnt building? because to me that seems like once this gets merged none of them will work.

**Comment by magnonellie** [comment]
@LucioFranco Currently I'm hoping that @ebkalderon will answer do that we can have the CI service be located on ci.amethyst.rs. This PR isn't building because the checkboxes aren't ticked off yet. As soon as I upload the new config to concourse, the bot gets write permissions and the webhook is in place, it's gonna start building all PRs.

**Comment by LucioFranco** [comment]
@magnonellie can we not add the bot as a collaborator to this repo?

File: README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 <img align="left" width="64px" src="book/src/images/amethyst_thumb.png" />
 
-# Amethyst
+# Godot
```
Comment (Moxinilian) [inline README.md]: I don't think you should include unsafe blocks like this. It's not really needed here.

**Comment by Moxinilian** [comment]
Is everybody okay with starting the bors process?

**Comment by torkleyy** [comment]
Like the bors try process or the one quickly (or not so quickly) merging
the commits into master?

me: fine with both ^^

**Comment by Moxinilian** [comment]
I was thinking about merging, but doing a try right now could be useful.
bors try

**Comment by Moxinilian** [comment]
It might be an isolated incident.
bors try
again

**Comment by Moxinilian** [comment]
Windows still failing for the same reason.

**Comment by torkleyy** [comment]
It's weird how it doesn't even finish the line..

**Comment by jojolepro** [comment]
bors try

**Comment by LucioFranco** [comment]
bors try

**Comment by Moxinilian** [comment]
Okay good.
Is everybody okay with the changes?
Is there anything left to do?

**Comment by torkleyy** [comment]
:shipit:

**Pull request merged**